hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/metadata/DatanodeTable.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.container.metadata;
+
+import org.apache.hadoop.hdds.utils.MetadataKeyFilters;
+import org.apache.hadoop.hdds.utils.db.Table;
+import org.apache.hadoop.hdds.utils.db.BatchOperation;
+import org.apache.hadoop.hdds.utils.db.cache.CacheKey;
+import org.apache.hadoop.hdds.utils.db.TableIterator;
+import org.apache.hadoop.hdds.utils.db.cache.CacheValue;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Wrapper class to represent a table in a datanode RocksDB instance.
+ * This class can wrap any {@link Table} instance, but will throw
+ * {@link UnsupportedOperationException} for {@link Table#iterator}.
+ * This is because differing schema versions used in datanode DB layouts may
+ * have differing underlying table structures, so iterating a table instance
+ * directly, without taking into account key prefixes, may yield unexpected
+ * results.
+ */
+public class DatanodeTable<KEY, VALUE> implements Table<KEY, VALUE> {
+
+  private final Table<KEY, VALUE> table;
+
```

Comment:
Because tables are now strongly typed, reads and writes to a table must have the same type, regardless of schema version. Having a different type for schema v1 and v2 tables will force callers to know which schema version they are working with, which we do not want. Therefore, since the value type of deleted block data was changed, all new reads and writes must use this type. This means the data written with the old schema version can no longer be read, but since it was just a duplicate of the key value (effectively a placeholder), this is not an issue.

Reply:
This would be nice, but unfortunately the AbstractDatanodeStore that is creating the DatanodeTables to return is getting pre-constructed instances of the Table interface from the DBColumnFamilyDefinition class. AbstractDatanodeStore must then convert these to DatanodeTables.

Lets say we make DatanodeTable extend TypedTable. Since AbstractDatanodeStore is given an existing instance that is not a DatanodeTable, creating a DatanodeTable from this requires assuming the original table is a TypedTable (which is currently true but may not hold in the future), casting it, and somehow copying all its internal state over to a new DatanodeTable.

Although using composition here requires a lot of wrapper methods that just call into the composed object, it is ultimately a cleaner approach because it allows creating a DatanodeTable from any existing Table implementation.